lib/Transforms/Scalar/PromoteMemcpy.cc

```diff
@@ -0,0 +1,211 @@
+
+private:
+  Module *m_M = nullptr;
+  LLVMContext *m_Ctx = nullptr;
+  const DataLayout *m_DL = nullptr;
+  DominatorTree *m_DT = nullptr;
+  AssumptionCache *m_AC = nullptr;
+
+  bool simplifyMemCpy(MemCpyInst *MCpy);
+};
+
+char PromoteMemcpy::ID = 0;
+
+bool PromoteMemcpy::simplifyMemCpy(MemCpyInst *MI) {
+  assert(MI);
+
+  unsigned DstAlign =
+      getKnownAlignment(MI->getArgOperand(0), *m_DL, MI, m_AC, m_DT);
+  unsigned SrcAlign =
+      getKnownAlignment(MI->getArgOperand(1), *m_DL, MI, m_AC, m_DT);
+  unsigned MinAlign = std::min(DstAlign, SrcAlign);
+  unsigned CopyAlign = MI->getAlignment();
+
+  if (CopyAlign < MinAlign) {
+    MI->setAlignment(ConstantInt::get(MI->getAlignmentType(), MinAlign, false));
+    return MI;
+  }
+
+  // If MemCpyInst length is 1/2/4/8 bytes then replace memcpy with
+  // load/store.
+  ConstantInt *MemOpLength = dyn_cast<ConstantInt>(MI->getArgOperand(2));
+  if (!MemOpLength)
+    return false;
+
+  // Source and destination pointer types are always "i8*" for intrinsic.  See
+  // if the size is something we can handle with a single primitive load/store.
+  // A single load+store correctly handles overlapping memory in the memmove
+  // case.
+  uint64_t Size = MemOpLength->getLimitedValue();
+  if (Size == 0)
+    return false;
+
+  auto *SrcPtr = MI->getArgOperand(1);
+  auto *DstPtr = MI->getArgOperand(0);
+
+  unsigned SrcAddrSp = cast<PointerType>(SrcPtr->getType())->getAddressSpace();
+  unsigned DstAddrSp = cast<PointerType>(DstPtr->getType())->getAddressSpace();
+
+  if (SrcAddrSp != DstAddrSp)
+    return false;
+
+  auto *SrcBitCast = dyn_cast<BitCastInst>(SrcPtr);
+  auto *DstBitCast = dyn_cast<BitCastInst>(DstPtr);
+
+  if (!SrcBitCast || !DstBitCast)
+    return false;
+
+  auto *Src = SrcBitCast->getOperand(0);
+  auto *Dst = DstBitCast->getOperand(0);
+
+  auto *SrcPtrTy = dyn_cast<PointerType>(Src->getType());
+  assert(SrcPtrTy);
+
+  auto *DstPtrTy = dyn_cast<PointerType>(Dst->getType());
+  assert(DstPtrTy);
+
+  if (SrcPtrTy != DstPtrTy)
+    return false;
+
+  auto *Buffer = dyn_cast<AllocaInst>(Dst);
+  if (!Buffer)
+    return false;
+
+  PMCPY_LOG(dbgs() << "\nSrc:\t");
+  PMCPY_LOG(Src->print(dbgs()));
+  PMCPY_LOG(SrcPtrTy->print(dbgs() << "\t"));
+  PMCPY_LOG(dbgs() << "\nDst:\t");
+  PMCPY_LOG(Dst->print(dbgs()));
+  PMCPY_LOG(DstPtrTy->print(dbgs() << "\t"));
+  PMCPY_LOG(dbgs() << "\n");
+  PMCPY_LOG(dbgs().flush());
+
+  if (!SrcPtrTy->getPointerElementType()->isFirstClassType()) {
+    PMCPY_LOG(dbgs() << "Not a first class type!\n");
+    return false;
+  }
+
+  auto *BufferTy = dyn_cast<StructType>(SrcPtrTy->getPointerElementType());
+  if (!BufferTy)
+    return false;
+
+  IRBuilder<> Builder(MI);
+  auto *I64Ty = IntegerType::getInt64Ty(*m_Ctx);
+  auto *NullInt = Constant::getNullValue(I64Ty);
+  auto *I32Ty = IntegerType::getInt32Ty(*m_Ctx);
+
+  for (unsigned i = 0, e = BufferTy->getStructNumElements(); i != e; ++i) {
```

Comment:
This will not recursively copy nested structs, right?
It is fine to cut corners in that way, but would be good to add a warning so that we know this is happening.

Reply:
Is a recursive copy really what we want to do here?

Reply:
if there are nested structs in llvm then yes